src/lib/tips.ts

```diff
@@ -79,43 +78,65 @@ export const TipsSchema = z.array(TipSchema)
-  }] | order(_updatedAt asc) {
-        _id,
-        _type,
-        _updatedAt,
-        _createdAt,
-        title,
-        state,
-        description,
-        eggheadRailsLessonId,
-        summary,
-        body,
-        'tags': softwareLibraries[] {
-          ...(library-> {
-            name,
-            'label': slug.current,
-            'http_url': url,
-            'image_url': image.url
-          }),
-        },
-        "videoResourceId": resources[@->._type == 'videoResource'][0]->_id,
-        "muxPlaybackId": resources[@->._type == 'videoResource'][0]-> muxAsset.muxPlaybackId,
-        "duration": resources[@->._type == 'videoResource'][0]->duration,
-        "slug": slug.current,
-        "transcript": resources[@->._type == 'videoResource'][0]-> castingwords.transcript,
-        "tweetId":  resources[@._type == 'tweet'][0].tweetId,
-        'instructor': collaborators[@->.role == 'instructor'][0]->{
-          title,
-          'slug': person->slug.current,
-          'name': person->name,
-          'path': person->website,
-          'twitter': person->twitter,
-          'image': person->image.url
-        },
-  }`)
+const GetAllTipsOptionsSchema = z.object({
+  onlyPublished: z.boolean(),
+  limit: z.number().optional(),
+})
+
+type GetAllTipsOptions = z.infer<typeof GetAllTipsOptionsSchema>
+
+export const getAllTips = async (
+  options: GetAllTipsOptions,
+): Promise<Tip[]> => {
+  const {onlyPublished, limit} = GetAllTipsOptionsSchema.parse(options)
+
+  // determine filtering conditions
+  const filterConditions = ['_type == "tip"']
+  if (onlyPublished) {
+    filterConditions.push('state == "published"')
+  }
+  const filter = filterConditions.join(' && ')
+
+  // determine slice clause limiting number of fetched results
+  const validLimit = limit && limit > 0
+  const sliceClause = validLimit ? `[0...${limit}]` : ''
+
+  const query = groq`*[${filter}] | order(_updatedAt desc) {
+    _id,
+    _type,
+    _updatedAt,
+    _createdAt,
+    title,
+    state,
+    description,
+    eggheadRailsLessonId,
+    summary,
+    body,
+    'tags': softwareLibraries[] {
+      ...(library-> {
+        name,
+        'label': slug.current,
+        'http_url': url,
+        'image_url': image.url
+      }),
+    },
+    "videoResourceId": resources[@->._type == 'videoResource'][0]->_id,
+    "muxPlaybackId": resources[@->._type == 'videoResource'][0]-> muxAsset.muxPlaybackId,
+    "duration": resources[@->._type == 'videoResource'][0]->duration,
+    "slug": slug.current,
+    "transcript": resources[@->._type == 'videoResource'][0]-> castingwords.transcript,
+    "tweetId":  resources[@._type == 'tweet'][0].tweetId,
+    'instructor': collaborators[@->.role == 'instructor'][0]->{
+      title,
+      'slug': person->slug.current,
+      'name': person->name,
+      'path': person->website,
+      'twitter': person->twitter,
+      'image': person->image.url
+    },
+  }${sliceClause}`
```

Comment:
Move the slice to the end because that seems to make results more consistent when also using the order by.